File: shared/src/iosTest/kotlin/com.commuto.interfacemobile.dbService/DBServiceTest.kt
package com.commuto.interfacemobile.dbService

import com.commuto.interfacemobile.db.DatabaseDriverFactory
import com.commuto.interfacemobile.db.KeyPair
import com.commuto.interfacemobile.db.PublicKey
import kotlin.test.*

class DBServiceTest {
    private var driver = DatabaseDriverFactory()
    private var dbService = DBService(driver)

    @BeforeTest
    fun testClearDatabase() {
        dbService.clearDatabase()
    }

    @Test
    fun testStoreAndGetKeyPair() {
        dbService.storeKeyPair("interf_id", "pub_key", "priv_key")
        val expectedKeyPair = KeyPair("interf_id", "pub_key", "priv_key")
        val keyPair: KeyPair? = dbService.getKeyPair("interf_id")
        assertEquals(keyPair!!, expectedKeyPair)
    }

    @Test
    fun testStoreAndGetPublicKey() {
        dbService.storePublicKey("interf_id", "pub_key")
        val expectedPublicKey = PublicKey("interf_id", "pub_key")
        val publicKey: PublicKey? = dbService.getPublicKey("interf_id")
        assertEquals(publicKey!!, expectedPublicKey)
    }

    @Test
    fun testDuplicateKeyPairProtection() {
        dbService.storeKeyPair("interf_id", "pub_key", "priv_key")
        try {
            dbService.storeKeyPair("interf_id", "pub_key", "priv_key")
        } catch (exception: IllegalStateException) {
            assert(exception.message!!.contains("Database query for key pair with interface id interf_id " +
                    "returned result"))
        }
    }

    @Test
    fun testDuplicatePublicKeyProtection() {
        dbService.storePublicKey("interf_id", "pub_key")
        try {
            dbService.storePublicKey("interf_id", "pub_key")
        } catch (exception: IllegalStateException) {
            assert(exception.message!!.contains("Database query for public key with interface id interf_id " +
                    "returned result"))
        }
    }
}

File: iosApp/iosAppTests/DBServiceTest.swift
//
//  DBServiceTest.swift
//  iosAppTests
//
//  Created by jimmyt on 11/28/21.
//  Copyright © 2021 orgName. All rights reserved.
//

import XCTest
@testable import iosApp

class DBServiceTest: XCTestCase {
    
    let dbService = DBService()

    override func setUpWithError() throws {
        // Put setup code here. This method is called before the invocation of each test method in the class.
        try dbService.connectToDb()
        try dbService.createTables()
    }

    override func tearDownWithError() throws {
        // Put teardown code here. This method is called after the invocation of each test method in the class.
    }
    
    func testStoreAndGetKeyPair() throws {
        try dbService.storeKeyPair(interfaceId: "interf_id", publicKey: "pub_key", privateKey: "priv_key")
        let expectedKeyPair = DBKeyPair(interfaceId: "interf_id", publicKey: "pub_key", privateKey: "priv_key")
        let keyPair: DBKeyPair? = try dbService.getKeyPair(interfaceId: "interf_id")
        XCTAssertEqual(expectedKeyPair, keyPair!)
    }
    
    func testStoreAndGetPublicKey() throws {
        try dbService.storePublicKey(interfaceId: "interf_id", publicKey: "pub_key")
        let expectedPublicKey = DBPublicKey(interfaceId: "interf_id", publicKey: "pub_key")
        let publicKey: DBPublicKey? = try dbService.getPublicKey(interfaceId: "interf_id")
        XCTAssertEqual(expectedPublicKey, publicKey!)
    }
    
    func testDuplicateKeyPairProtection() throws {
        try dbService.storeKeyPair(interfaceId: "interf_id", publicKey: "pub_key", privateKey: "priv_key")
        do {
            try dbService.storeKeyPair(interfaceId: "interf_id", publicKey: "pub_key", privateKey: "priv_key")
        } catch DBService.DBServiceError.unexpectedDbResult(let message) where message == "Database query for key pair with interface id interf_id returned result" {
            
        }
    }
    
    func testDuplicatePublicKeyProtection() throws {
        try dbService.storePublicKey(interfaceId: "interf_id", publicKey: "pub_key")
        do {
            try dbService.storePublicKey(interfaceId: "interf_id", publicKey: "pub_key")
        } catch DBService.DBServiceError.unexpectedDbResult(let message) where message == "Database query for public key with interface id interf_id returned result" {
            
        }
    }

}
